File: Searching/704_Binary_Search/solution704.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            m = l + ((r - l) // 2)
            if nums[m] > target:
                r = m - 1
            elif nums[m] < target:
                l = m + 1
            else:
                return m
        return -1

# Runtime1: 189ms, beats 67.96%
# Memory1: 18.10MB, beats 71.59%

# Runtime2: 188ms, beats 73.18%
# Memory2: 18.09MB, beats 71.59%

# Runtime3: 193ms, beats 43.76%
# Memory3: 18.07MB, beats 71.59%

File: Searching/704_Binary_Search/solution704.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        
        while (l <= r) {
            int m = l + (r - l) / 2;
            
            if (nums[m] > target) {
                r = m - 1;
            } else if (nums[m] < target) {
                l = m + 1;
            } else {
                return m;
            }
        }
        
        return -1;
    }
};

// Time Complexity:
// Space Complexity:

// Runtime1: 22ms, beats 86.36%
// Memory1: 30.22MB, beats 23.48%

// Runtime2: 26ms, beats 62.49%
// Memory2: 30.27MB, beats 23.48%

// Runtime3: 15ms, beats 98.50%
// Memory3: 30.20MB, beats 73.73%